### PR TITLE
Add method matching to HTTPRoute matching precedence order

### PR DESCRIPTION
**Pull request by gauravkghildiyal**
**What type of PR is this?**
/kind cleanup
/kind documentation
/kind test
/area conformance

**What this PR does / why we need it**:
Clarify the precedency that Method matching has while comparing similar HTTPRouteRules.

**Which issue(s) this PR fixes**:

Fixes https://github.com/kubernetes-sigs/gateway-api/issues/1993

**Does this PR introduce a user-facing change?**:

```release-note
HTTPRoute Method matching precedence has been clarified
```

### CODE REVIEW

**Comment by robscott** [comment]
Thanks @gauravkghildiyal!

/ok-to-test
/cc @arkodg @mlavacca @sunjayBhatia

File: apis/v1beta1/httproute_types.go [hunk 1]
```diff
@@ -164,6 +164,7 @@ type HTTPRouteRule struct {
 	// * Characters in a matching "Prefix" path match
 	// * Header matches.
 	// * Query param matches.
+	// * Method matches.
```
Comment (arkodg) [inline apis/v1beta1/httproute_types.go:167]: curious why its after `Query param` ? is it to maintain backwards compatibility ?

Comment (howardjohn) [reply apis/v1beta1/httproute_types.go:167]: +1. I would put it before "Headers" personally

Comment (sunjayBhatia) [reply apis/v1beta1/httproute_types.go:167]: (also seems like the language before this list needs to be amended, currently it reads `precedence must be given to the match with the largest number of:`, and adding method matching in here doesn't really fit)

Comment (robscott) [reply apis/v1beta1/httproute_types.go:167]: @gauravkghildiyal at today's meeting I think we had consensus that this should fit right above header matches. Can you update that and the corresponding conformance test?

Comment (gauravkghildiyal) [reply apis/v1beta1/httproute_types.go]: Thanks for the info. Updated!

File: apis/v1beta1/httproute_types.go [hunk 2]
```diff
@@ -158,12 +158,13 @@ type HTTPRouteRule struct {
 	// Proxy or Load Balancer routing configuration generated from HTTPRoutes
 	// MUST prioritize matches based on the following criteria, continuing on
 	// ties. Across all rules specified on applicable Routes, precedence must be
-	// given to the match with the largest number of:
+	// given to the rule having:
 	//
-	// * Characters in a matching "Exact" path match
```
Comment (gauravkghildiyal) [inline apis/v1beta1/httproute_types.go:163]: Please pay attention to the rewrite here -- I felt the "number of characters in exact path match" may not have been entirely appropriate. Please do point out if that had some relevance, which I'm missing, and I'll correct the statements.

Comment (mlavacca) [reply apis/v1beta1/httproute_types.go:163]: I think that this re-writing related to the exact path match is more accurate 👍

File: apis/v1beta1/httproute_types.go [hunk 3]
```diff
@@ -158,12 +158,13 @@ type HTTPRouteRule struct {
 	// Proxy or Load Balancer routing configuration generated from HTTPRoutes
 	// MUST prioritize matches based on the following criteria, continuing on
 	// ties. Across all rules specified on applicable Routes, precedence must be
-	// given to the match with the largest number of:
+	// given to the rule having:
```
Comment (mlavacca) [inline apis/v1beta1/httproute_types.go:161]: I think that `given to the match having` is more accurate than `given to the rule having` wording, as the field specifying the `HTTPRouteMatch` is a match actually, not a rule.

Comment (gauravkghildiyal) [reply apis/v1beta1/httproute_types.go]: Thanks @mlavacca. Yes that is indeed more appropriate.

**Review by mlavacca — APPROVED** [review]
Thanks @gauravkghildiyal!

**Review by sunjayBhatia — APPROVED** [review]

File: apis/v1beta1/httproute_types.go (hunk 2 above)
Comment (youngnick) [reply apis/v1beta1/httproute_types.go:163]: The reason for the "number of characters" part is to handle the case of two rules with Exact path matching - so the longer one will win in that case.

Comment (youngnick) [reply apis/v1beta1/httproute_types.go:163]: Ah right, I think I missed that nuance. I withdraw my objection. 😄

**Comment by robscott** [comment]
Thanks @gauravkghildiyal!

/lgtm
/approve
/cherry-pick release-0.7

**Pull request merged**